File: ulib/ml/FStar_UInt64.ml
type uint64 = Stdint.Uint64.t
type uint8 = int
type t = uint64
type t' = t
           
let (%) x y = if x < 0 then (x mod y) + y else x mod y

let n = Prims.parse_int "64"
let v (x:uint64) : Prims.int = Prims.parse_int (Stdint.Uint64.to_string x)

let zero = Stdint.Uint64.zero
let one = Stdint.Uint64.one
let ones = Stdint.Uint64.pred Stdint.Uint64.zero

let add (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.add a b
let add_underspec a b = add a b
let add_mod a b = add a b

let sub (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.sub a b
let sub_underspec a b = sub a b
let sub_mod a b = sub a b

let mul (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.mul a b
let mul_underspec a b = mul a b
let mul_mod a b = mul a b
let mul_div a b =
  let a = Stdint.Uint128.of_uint64 a in
  let b = Stdint.Uint128.of_uint64 b in
  let r = Stdint.Uint128.mul a b in
  let r = Stdint.Uint128.shift_right r 64 in
  Stdint.Uint64.of_uint128 r

let div (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.div a b

let rem (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.rem a b

let logand (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.logand a b
let logxor (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.logxor a b
let logor  (a:uint64) (b:uint64) : uint64 = Stdint.Uint64.logor a b
let lognot (a:uint64) : uint64 = Stdint.Uint64.lognot a
                                              
let int_to_uint64 (x:Prims.int) : uint64= Stdint.Uint64.of_string (Prims.to_string x)

let shift_right (a:uint64) (b:uint8) : uint64 = Stdint.Uint64.shift_right a b
let shift_left  (a:uint64) (b:uint8) : uint64 = Stdint.Uint64.shift_left a b

(* Comparison operators *)
let eq (a:uint64) (b:uint64) : bool = a = b
let gt (a:uint64) (b:uint64) : bool = a > b
let gte (a:uint64) (b:uint64) : bool = a >= b
let lt (a:uint64) (b:uint64) : bool = a < b
let lte (a:uint64) (b:uint64) : bool =  a <= b

(* NOT Constant time operators *)
let eq_mask (a:uint64) (b:uint64) : uint64 = if a = b then Stdint.Uint64.pred Stdint.Uint64.zero
                                             else Stdint.Uint64.zero
let gte_mask (a:uint64) (b:uint64) : uint64 = if a >= b then Stdint.Uint64.pred Stdint.Uint64.zero
                                             else Stdint.Uint64.zero
                                               
(* Infix notations *)
let op_Plus_Hat = add
let op_Plus_Question_Hat = add_underspec
let op_Plus_Percent_Hat = add_mod
let op_Subtraction_Hat = sub
let op_Subtraction_Question_Hat = sub_underspec
let op_Subtraction_Percent_Hat = sub_mod
let op_Star_Hat = mul
let op_Star_Question_Hat = mul_underspec
let op_Star_Percent_Hat = mul_mod
let op_Slash_Hat = div
let op_Percent_Hat = rem
let op_Hat_Hat = logxor  
let op_Amp_Hat = logand
let op_Bar_Hat = logor
let op_Less_Less_Hat = shift_left
let op_Greater_Greater_Hat = shift_right
let op_Equals_Hat = eq
let op_Greater_Hat = gt
let op_Greater_Equals_Hat = gte
let op_Less_Hat = lt
let op_Less_Equals_Hat = lte

let to_string s = Stdint.Uint64.to_string s
let of_string s = Stdint.Uint64.of_string s

let uint_to_t s = Stdint.Uint64.of_string (Z.to_string s)
let __uint_to_t = uint_to_t

File: examples/javascript/poly1305_js/FStar_UInt64.js
/* @flow */
import * as FStar_UInt32 from "./FStar_UInt32";
import * as FStar_UInt from "./FStar_UInt";
export const n: number = 64;

export type Mk = {_tag: "Mk", _0: FStar_UInt.uint_t<null>};

export type t_ = Mk;

const uu___is_Mk = (projectee: t_): (boolean) => (true);
const __proj__Mk__item__v = (projectee: t_): (FStar_UInt.uint_t<null>) => {
  let _res;
  const _0_112 = projectee;
  if( (_0_112._tag === "Mk") ) {
    const v = _0_112._0;
    _res = v;
  } else {
    throw "This value doesn\'t match!";
  }
  return _res;
};
export type t = t_;

export const v = (x: t): (FStar_UInt.uint_t<null>) => (__proj__Mk__item__v(x));

export const add = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.add(n)(v(a))(v(b))
}));

export const add_underspec = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.add_underspec(n)(v(a))(v(b))
}));

export const add_mod = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.add_mod(n)(v(a))(v(b))
}));

export const sub = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.sub(n)(v(a))(v(b))
}));

export const sub_underspec = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.sub_underspec(n)(v(a))(v(b))
}));

export const sub_mod = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.sub_mod(n)(v(a))(v(b))
}));

export const mul = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.mul(n)(v(a))(v(b))
}));

export const mul_underspec = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.mul_underspec(n)(v(a))(v(b))
}));

export const mul_mod = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.mul_mod(n)(v(a))(v(b))
}));

export const div = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.div(n)(v(a))(v(b))
}));

export const div_underspec = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.div_underspec(n)(v(a))(v(b))
}));

export const rem = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.mod_(n)(v(a))(v(b))
}));

export const logand = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.logand(n)(v(a))(v(b))
}));

export const logxor = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.logxor(n)(v(a))(v(b))
}));

export const logor = (a: t): ((_1: t) => (t)) => ((b: t) => ({
  _tag: "Mk",
  _0: FStar_UInt.logor(n)(v(a))(v(b))
}));

export const lognot = (a: t): (t) => ({
  _tag: "Mk",
  _0: FStar_UInt.lognot(n)(v(a))
});

export const uint_to_t = (x: FStar_UInt.uint_t<null> ): (t) => ({
  _tag: "Mk",
  _0: x
});

export const shift_right = (a: t): ((_1: FStar_UInt32.t) => (t)) => ((s: FStar_UInt32.t) => ({
  _tag: "Mk",
  _0: FStar_UInt.shift_right(n)(v(a))(FStar_UInt32.v(s))
}));

export const shift_left = (a: t): ((_1: FStar_UInt32.t) => (t)) => ((s: FStar_UInt32.t) => ({
  _tag: "Mk",
  _0: FStar_UInt.shift_left(n)(v(a))(FStar_UInt32.v(s))
}));

export const eq = (a: t): ((_1: t) => (boolean)) => ((b: t) => (FStar_UInt.eq(n)(v(a))(v(b))));

export const gt = (a: t): ((_1: t) => (boolean)) => ((b: t) => (FStar_UInt.gt(n)(v(a))(v(b))));

export const gte = (a: t): ((_1: t) => (boolean)) => ((b: t) => (FStar_UInt.gte(n)(v(a))(v(b))));

export const lt = (a: t): ((_1: t) => (boolean)) => ((b: t) => (FStar_UInt.lt(n)(v(a))(v(b))));

export const lte = (a: t): ((_1: t) => (boolean)) => ((b: t) => (FStar_UInt.lte(n)(v(a))(v(b))));

export const eq_mask = (a: t) => ((b: t) => {
  throw "Not yet implemented!";
});

export const gte_mask = (a: t) => ((b: t) => {
  throw "Not yet implemented!";
});

export const op_Plus_Hat: (_1: t) => ((_1: t) => (t)) = add;

export const op_Plus_Question_Hat: (_1: t) => ((_1: t) => (t)) = add_underspec;

export const op_Plus_Percent_Hat: (_1: t) => ((_1: t) => (t)) = add_mod;

export const op_Subtraction_Hat: (_1: t) => ((_1: t) => (t)) = sub;

export const op_Subtraction_Question_Hat: (_1: t) => ((_1: t) => (t)) = sub_underspec;

export const op_Subtraction_Percent_Hat: (_1: t) => ((_1: t) => (t)) = sub_mod;

export const op_Star_Hat: (_1: t) => ((_1: t) => (t)) = mul;

export const op_Star_Question_Hat: (_1: t) => ((_1: t) => (t)) = mul_underspec;

export const op_Star_Percent_Hat: (_1: t) => ((_1: t) => (t)) = mul_mod;

export const op_Slash_Hat: (_1: t) => ((_1: t) => (t)) = div;

export const op_Percent_Hat: (_1: t) => ((_1: t) => (t)) = rem;

export const op_Hat_Hat: (_1: t) => ((_1: t) => (t)) = logxor;

export const op_Amp_Hat: (_1: t) => ((_1: t) => (t)) = logand;

export const op_Bar_Hat: (_1: t) => ((_1: t) => (t)) = logor;

export const op_Less_Less_Hat: (_1: t) => ((_1: FStar_UInt32.t) => (t)) = shift_left;

export const op_Greater_Greater_Hat: (_1: t) => ((_1: FStar_UInt32.t) => (t)) = shift_right;

export const op_Equals_Hat: (_1: t) => ((_1: t) => (boolean)) = eq;

export const op_Greater_Hat: (_1: t) => ((_1: t) => (boolean)) = gt;

export const op_Greater_Equals_Hat: (_1: t) => ((_1: t) => (boolean)) = gte;

export const op_Less_Hat: (_1: t) => ((_1: t) => (boolean)) = lt;

export const op_Less_Equals_Hat: (_1: t) => ((_1: t) => (boolean)) = lte;

export const to_string = (uu____999: t) => {
  throw "Not yet implemented!";
};

export const of_string = (uu____1004: string) => {
  throw "Not yet implemented!";
};


